### ISSUE TITLE
autocompletes

### ISSUE THREAD

**Issue by kobbejager** [CarlosNZ/fig-tree-evaluator#121]
Hi

Looking into jsonforms-with-figtree-demo and this repo, I was wondering if fig-tree-evaluator can be used to build an autocomplete renderer in JSON Forms? In other words, do you think it is possible to dynamically (re-)run $GET as a user types?

**Comment by CarlosNZ** [comment]
Yes, this would work fine.

You can see a quick-and-dirty version of this in action in the [demo](https://github.com/CarlosNZ/jsonforms-with-figtree-demo) by changing the "Country" input from a drop-down selector to a free text:

In the `schema.json`, change this:
```json
    "country": {
      "type": "string",
      "title": "Select your country",
      "enum": {
        "$GET": { "url": "https://countriesnow.space/api/v0.1/countries/flag/unicode", "returnProperty": "data.name" }
      }
    },
```

to this:
```json
    "country": {
      "type": "string",
      "title": "Select your country"
    },
```

Then you'll see that the "City" dropdown makes additional GET requests to get the city list as you type in your country.

This is very crude -- you'd need to do some better error handling to make it nice (it gives a bad UI if the country input is invalid, which it is until you finish typing a real country), but it gives an example of what you're asking I think. Obviously it's not a good real-life example (the drop-down country selector makes much more sense), but I hope it answers your question?

**Comment by kobbejager** [comment]
Thanks for your quick reply. 
This is not exactly what I mean. What you propose is filling an enum list of a field based on the input of another field. 

Imagine that I only have a [cities API endpoint with over 100k cities worldwide]('https://www.thecompaniesapi.com/api/search-cities'), and no country field. If I start typing character for character in this field, it queries the endpoint multiple times. E.g.

- 'b' -> https://api.thecompaniesapi.com/v2/locations/cities?search=b
- 'be' -> https://api.thecompaniesapi.com/v2/locations/cities?search=be
- 'ber' -> https://api.thecompaniesapi.com/v2/locations/cities?search=ber
- ... (until I can select Berlin from the dynamically updates dropdown list)

**Comment by CarlosNZ** [comment]
Right, yes, I see. It can be done, but it requires a [custom renderer](https://jsonforms.io/docs/tutorial/custom-renderers#1-add-the-control-to-uischemajson) for JSON Forms to make an AutoComplete component. I whipped one up using MUI and got it working in the way you describe. You can see it in action on the demo page now (right at the bottom): https://carlosnz.github.io/jsonforms-with-figtree-demo/

I had to do one slightly awkward thing in that the component writes a temporary value to the JSON Forms data from the user input, which is then passed in to the options query. With the way I've set up Fig Tree to work with JSON Forms, the FigTree evaluator doesn't have access to any state within components, only the JSON data itself, so have to write the search text value to another schema field on input. You can see this value labelled as `globalCitySearchText` in my example, but this value is removed once the city is selected.

Here's how it looks, as you type input:

<img width="429" alt="Screenshot 2024-11-25 at 1 05 46 AM" src="https://github.com/user-attachments/assets/3aeed227-6ca7-47f0-b320-760bd14eda7f">

And in the JSON data, before selection:

<img width="482" alt="Screenshot 2024-11-25 at 1 05 39 AM" src="https://github.com/user-attachments/assets/71ce88a4-3359-4da4-bf8d-820a400990f3">

Then, once selected:

<img width="443" alt="Screenshot 2024-11-25 at 1 05 56 AM" src="https://github.com/user-attachments/assets/ecb54fee-ae9a-4dc1-a0ee-6bf5f61e69c9">

And the JSON data after selection:

<img width="473" alt="Screenshot 2024-11-25 at 1 06 02 AM" src="https://github.com/user-attachments/assets/46fce588-e057-4a5f-9dda-b481ff42e0bf">

It has basic debouncing to prevent hammering the server. It's a little rough around the edges, but seems to be doing the job for a quick demo.

Thanks for your question -- it's good to see that this can work.
